### ISSUE TITLE
Wrapper component lifecycle issue

### ISSUE THREAD

**Issue by idolizesc** [tomchentw/react-google-maps#47]
The `GoogleMaps` wrapper component breaks the React lifecycle because it does not re-render the VirtualContainer until [after it is done updating everything](https://github.com/tomchentw/react-google-maps/blob/master/src/GoogleMaps.js#L79).

This means trying to get an updated value from a newly-changed `Marker`, for example, is not possible in any parent component's `componentDidUpdate` method, because the children of the VirtualContainer are not yet re-rendered.

This is a regression from the previous mixin/context methodology, and leads to very confusing and difficult to debug errors.

Perhaps one solution would be to re-render VirtualContainer on `componentWillUpdate` instead.

**Comment by tomchentw** [comment]
I see. Could you show me the reason why you'd like to read properties from a newly-changed `Marker`? Since if everything is rendered with given props, you already know the value from your data, right?

**Comment by idolize** [comment]
Well, my use case actually involves reading shape data from a component that renders `Circle` items. I need to access Google Maps API features, like `circle.getBounds()`, after some event was triggered.

I suppose I could create a `new google.maps.Circle` every time from the original input props, but that is annoying and less performant. Also I just feel like it's a reasonable exception for the components to stick to the standard React lifecycle and expose the correct data at that time via the getters.

**Comment by tomchentw** [comment]
#36

**Comment by tomchentw** [comment]
Just tried with your use case with [react-google-maps/#basics/geolocation](https://tomchentw.github.io/react-google-maps/#basics/geolocation), yes, changing callback to `componentWillUpdate` of `GoogleMaps` will fix this issue.

However, I'm thinking that the same use case may apply for `componentDidMount` since we do `_render_virtual_container_` in `GoogleMaps#componentDidMount` as well. As a result, I suspect changing to `componentWillUpdate` may not be a permanent solution.

**Comment by tomchentw** [comment]
The `VirtualContainer` technique prevents the leak of two wrapper `div`s around the `google.maps.Map` instance and react components. If we only use one wrapper `div`, then initialise  `google.maps.Map` instance on it as well as rendering children components inside it. When next update ticks, React will clean out everything created by `google.maps.Map` instance. Thus only `<noscript>` tags left inside the `<div>`

**Comment by idolize** [comment]
Interesting. React definitely needs some kind of first-class "fragment" element that does not render anything to the DOM. I know [they are working on it](https://groups.google.com/forum/#!topic/reactjs/pHNJe8trFOg).

In the mean time, have you looked at [react-outlet](https://github.com/carlsverre/react-outlet)? It could possibly be of interest here.

Another interesting approach is [the way MartyJS creates "container" components](https://github.com/martyjs/marty/blob/master/lib/createContainer/createContainer.js): calling `Marty.createContainer(SomeComponentToWrap)`, but also giving you the option to [override the method which actually renders the component](http://martyjs.org/api/containers/index.html#done).

**Comment by tomchentw** [comment]
https://github.com/facebook/react/issues/2191

**Comment by idolize** [comment]
Can we just ditch the extra `div` [being created to house the container](https://github.com/tomchentw/react-google-maps/blob/master/src/GoogleMaps.js#L79) and instead render the children as either children or siblings of the Google Maps div itself?

This would solve this issue as well as the issue where we can't render any DOM elements as children of the `GoogleMaps` component.

Also, looking more closely at the code, I believe there is a memory leak in the current implementation, as the `containerNode` is [never actually destroyed on unmount](https://github.com/tomchentw/react-google-maps/blob/master/src/GoogleMaps.js#L90) (just set to null), so new container `divs` will be created each time the component is mounted.

**Comment by tomchentw** [comment]
@idolize , could you please provide an example of `we can't render any DOM elements as children of the GoogleMaps component`. Cause I cannot understand why you'd like to do it this way. A simple gist would be enough.

Thanks!

**Comment by idolize** [comment]
Yeah, my use case is I want to use my own DOM to replace some Google Maps controls. The Google Maps API allows you to do this for almost any control (such as the zoom indicator, panning buttons, drawing controls, etc.)

Here is an example of how I could overwrite the drawing controls (which worked in the pre-wrapper component versions):

```js
import React from 'react';
import {DrawingManager} from 'react-google-maps';

export default React.createClass({
  getInitialState() {
    return { drawingMode: null };
  },
  setDrawingMode(drawingMode) {
    this.setState({ drawingMode });
  },
  render () {
    return (
      <div>
        <DrawingManager
          {...this.props}
          drawingControl={false} {/* Use our own buttons instead of the Google Maps built-in drawing control */}
          drawingMode={this.state.drawingMode}
        />
        <button onClick={this.setDrawingMode.bind(this, google.maps.drawing.OverlayType.CIRCLE)}>Circle</button>
        <button onClick={this.setDrawingMode.bind(this, google.maps.drawing.OverlayType.POLYGON)}>Polygon</button>
      </div>
    );
  }
});
```

But this doesn't work with the current implementation because the new drawing controls I try to render don't show up, as they are rendered inside the virtual container div.

**Comment by tomchentw** [comment]
I believe this was fixed in #61. And I certainly believe that I use a elegant way for component lifecycle in the rewrite of #88. Please take a look at it and comments are welcome, thanks!

**Comment by tomchentw** [comment]
Released [v2.0.0] and [v2.0.1].

[v2.0.0]: https://github.com/tomchentw/react-google-maps/releases/tag/v2.0.0
[v2.0.1]: https://github.com/tomchentw/react-google-maps/releases/tag/v2.0.1
